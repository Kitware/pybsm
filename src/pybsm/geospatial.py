"""The Python Based Sensor Model (pyBSM) is a collection of electro-optical camera modeling functions.

Developed by the Air Force Research Laboratory, Sensors Directorate.

Author citation:
LeMaster, Daniel A.; Eismann, Michael T., "pyBSM: A Python package for modeling
imaging systems", Proc. SPIE 10204 (2017)

Distribution A.  Approved for public release.
Public release approval for version 0.0: 88ABW-2017-3101
Public release approval for version 0.1: 88ABW-2018-5226

Maintainer: Kitware, Inc. <nrtk@kitware.com>
"""

# standard library imports
import inspect
import os
import warnings

# 3rd party imports
import numpy as np

# new in version 0.2.  We filter warnings associated with calculations in the
# function circularApertureOTF.  These invalid values are caught as NaNs and
# appropriately replaced.
warnings.filterwarnings("ignore", r"invalid value encountered in arccos")
warnings.filterwarnings("ignore", r"invalid value encountered in sqrt")
warnings.filterwarnings("ignore", r"invalid value encountered in true_divide")
warnings.filterwarnings("ignore", r"divide by zero encountered in true_divide")

# find the current path (used to locate the atmosphere database)
# dir_path = os.path.dirname(os.path.abspath(__file__))
dir_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))

# define some useful physical constants
r_earth = 6378.164e3  # radius of the earth (m)


def altitude_along_slant_path(
    h_target: float,
    h_sensor: float,
    slant_range: float,
) -> tuple[np.ndarray, np.ndarray]:
    """Calculate the height above the curved earth at points along a path from the target to the sensor.

    Calculate the height above the curved earth at points along a path from the
    target (z_path=0) to the sensor (z_path.max()).  This is primarily useful for
    calculating the atmospheric coherence diameter, r0.

    :param h_target:
        height of the target above sea level (m)
    :param h_sensor:
        height of the sensor above sea level (m)
    :param slant_range:
        distance between the target and sensor (m)

    :return:
        z_path:
            array of samples along the path from the target (z_path = 0) to the
            sensor (m)
        h_path:
            height above the earth along a slantpath defined by z_path (m)

    :raises:
        ZeroDivisionError:
            if slant_range is 0

    """
    # this is simple law of cosines problem
    nadir = nadir_angle(h_target, h_sensor, slant_range)
    a = r_earth + h_sensor
    b = np.linspace(0.0, slant_range, 10000)  # arbitrary choice of 100,000 data points
    c = np.sqrt(a**2 + b**2 - 2 * a * b * np.cos(nadir))

    z_path = b
    h_path = (c - r_earth)[::-1]
    # It is correct to reverse the order of h_path.  The height of the target
    # above the earth (h_target) should be the first element in the array.
    # Effectively, this reversal just changes the location of the origin of
    # z_path to the target location.  It was just more convenient to use the
    # origin at the sensor in the law of cosines calculation.

    return (z_path, h_path)


def ground_sample_distance(ifov: float, slant_range: float) -> float:
    """IBSM Equation 3-62, the ground sample distance.

    The ground sample distance, i.e. the footprint of a single detector in object space.

    :param ifov:
        instantaneous field-of-view of a detector (radians)
    :param slant_range:
        slant range to target (m)

    :return:
        gsd:
            ground sample distance (m)
    """
    return slant_range * ifov


def nadir_angle(h_target: float, h_sensor: float, slant_range: float) -> float:
    """Work through the law of cosines to calculate the sensor nadir angle above a circular earth.

    Work through the law of cosines to calculate the sensor nadir angle above a
    circular earth (i.e. angle between looking straight down (nadir = 0) and looking
    along the slant path).

    :param h_target:
        height of the target above sea level (m)
    :param h_sensor:
        height of the sensor above sea level (m)
    :param slant_range:
        distance between the target and sensor (m)

    :return:
        nadir:
            the sensor nadir angle (rad)

    :raises:
        ZeroDivisionError:
            if slant_range is 0
    """
    a = r_earth + h_sensor
    b = slant_range
    c = r_earth + h_target

    return np.arccos(-1.0 * (c**2.0 - a**2.0 - b**2.0) / (2.0 * a * b))


def curved_earth_slant_range(
    h_target: float,
    h_sensor: float,
    ground_range: float,
) -> float:
    """Returns the slant range from target to sensor above a curved (circular) Earth.

    :param h_target:
        height of the target above sea level (m)
    :param h_sensor:
        height of the sensor above sea level (m)
    :param ground_range:
        distance between the target and sensor on the ground (m)

    :return:
        slant_range:
            distance between the target and sensor (m)
    """
    a = r_earth + h_sensor
    c = r_earth + h_target
    theta = ground_range / r_earth  # exact arc length angle (easy to derive)

    return np.sqrt(c**2.0 + a**2.0 - 2.0 * a * c * np.cos(theta))
