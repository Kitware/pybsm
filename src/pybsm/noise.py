"""The Python Based Sensor Model (pyBSM) is a collection of electro-optical camera modeling functions.

Developed by the Air Force Research Laboratory, Sensors Directorate.

Author citation:
LeMaster, Daniel A.; Eismann, Michael T., "pyBSM: A Python package for modeling
imaging systems", Proc. SPIE 10204 (2017)

Distribution A.  Approved for public release.
Public release approval for version 0.0: 88ABW-2017-3101
Public release approval for version 0.1: 88ABW-2018-5226

Maintainer: Kitware, Inc. <nrtk@kitware.com>
"""

# standard library imports
import inspect
import os
import warnings

# 3rd party imports
import numpy as np

# new in version 0.2.  We filter warnings associated with calculations in the
# function circularApertureOTF.  These invalid values are caught as NaNs and
# appropriately replaced.
warnings.filterwarnings("ignore", r"invalid value encountered in arccos")
warnings.filterwarnings("ignore", r"invalid value encountered in sqrt")
warnings.filterwarnings("ignore", r"invalid value encountered in true_divide")
warnings.filterwarnings("ignore", r"divide by zero encountered in true_divide")

# find the current path (used to locate the atmosphere database)
# dir_path = os.path.dirname(os.path.abspath(__file__))
dir_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))


def noise_gain(kernel: np.ndarray) -> float:
    """Noise Gain is the GIQE term representing increase in noise due to image sharpening.

    The definition is not included in the IBSM manual.  This version comes from
    Leachtenauer et al., "General Image-Quality Equation: GIQE" APPLIED OPTICS
    Vol. 36, No. 32 10 November 1997.

    :param kernel:
         the 2D image sharpening kernel; note that
         the kernel is assumed to sum to one

    :return:
        ng:
            noise gain (unitless)

    :raises:
        ValueError:
            if values in kernel do not sum to 1
    """
    if np.sum(kernel) != 1.0:
        raise ValueError("Kernel does not sum to 1")

    return np.sqrt(np.sum(np.sum(kernel**2)))


def quantization_noise(pe_range: float, bit_depth: float) -> float:
    """Effective noise contribution from the number of photoelectrons quantized.

    Effective noise contribution from the number of photoelectrons quantized by a single count of the analog
    to digital converter.Quantization noise is buried in the definition of signal-to-noise in IBSM equation 3-47.

    :param pe_range:
        the difference between the maximum and minimum number of photoelectrons
        that may be sampled by the readout electronics (e-)
    :param bit_depth:
        number of bits in the analog to digital converter (unitless)

    :return:
        sigma_q :
            quantization noise given as a photoelectron standard deviation (e-)

    :WARNING:
        output can be nan if pe_range is 0
        output can be inf if bit_depth is 0
    """
    return pe_range / (np.sqrt(12) * (2.0**bit_depth - 1.0))
