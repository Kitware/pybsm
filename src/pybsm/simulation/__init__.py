from .sensor import Sensor  # noqa F401
from .scenario import Scenario  # noqa F401
from .ref_image import RefImage  # noqa F401
from .functional import *  # noqa F401,F403
from .image_simulator import *  # noqa F401,F403
