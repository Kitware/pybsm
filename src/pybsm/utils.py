"""The Python Based Sensor Model (pyBSM) is a collection of electro-optical camera modeling functions.

Developed by the Air Force Research Laboratory, Sensors Directorate.

Author citation:
LeMaster, Daniel A.; Eismann, Michael T., "pyBSM: A Python package for modeling
imaging systems", Proc. SPIE 10204 (2017)

Distribution A.  Approved for public release.
Public release approval for version 0.0: 88ABW-2017-3101
Public release approval for version 0.1: 88ABW-2018-5226

Maintainer: Kitware, Inc. <nrtk@kitware.com>
"""

# standard library imports
import inspect
import os
import warnings

# 3rd party imports
import numpy as np

# new in version 0.2.  We filter warnings associated with calculations in the
# function circularApertureOTF.  These invalid values are caught as NaNs and
# appropriately replaced.
warnings.filterwarnings("ignore", r"invalid value encountered in arccos")
warnings.filterwarnings("ignore", r"invalid value encountered in sqrt")
warnings.filterwarnings("ignore", r"invalid value encountered in true_divide")
warnings.filterwarnings("ignore", r"divide by zero encountered in true_divide")

# find the current path (used to locate the atmosphere database)
# dir_path = os.path.dirname(os.path.abspath(__file__))
dir_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))


def load_database_atmosphere_no_interp(
    altitude: float,
    ground_range: float,
    ihaze: int,
) -> np.ndarray:
    """Loads a precalculated MODTRAN 5.2.1 Tape 7 atmosphere over a wavelength range of 0.3 to 14 micrometers.

    All screnario details are in 'atmosphere_README.txt'. NOTE: the _no_interp suffix was added for version
    0.2. See pybsm.load_database_atmosphere for more information.

    :param altitude:
        sensor height above ground level in meters.  The database includes the
        following altitude options: 2 32.55 75 150 225 500 meters, 1000 to
        12000 in 1000 meter steps, and 14000 to 20000 in 2000 meter steps,
        24500 meters
    :param ground_range:
        distance *on the ground* between the target and sensor in meters.
        The following ground ranges are included in the database at each
        altitude until the ground range exceeds the distance to the spherical
        earth horizon: 0 100 500 1000 to 20000 in 1000 meter steps, 22000 to
        80000 in 2000 m steps, and  85000 to 300000 in 5000 meter steps.
    :param ihaze:
        MODTRAN code for visibility, valid options are ihaze = 1 (Rural
        extinction with 23 km visibility) or ihaze = 2 (Rural extinction
        with 5 km visibility)


    :return:
        atm[:,0]:
            wavelengths from .3 to 14 x 10^-6 m in 0.01x10^-6 m steps
        atm[:,1]:
            (TRANS) total transmission through the defined path.
        atm[:,2]:
            (PTH THRML) radiance component due to atmospheric emission and
            scattering received at the observer.
        atm[:,3]:
            (SURF EMIS) component of radiance due to surface emission received
            at the observer.
        atm[:,4]:
            (SOL SCAT) component of scattered solar radiance received at the
            observer.
        atm[:,5]:
            (GRND RFLT) is the total solar flux impingent on the ground and
            reflected directly to the sensor from the ground. (direct radiance
            + diffuse radiance) * surface reflectance
    :NOTE: units for columns 1 through 5 are in radiance W/(sr m^2 m)
    """
    # decoder maps filenames to atmospheric attributes
    atm_path = os.path.join(dir_path, "atms", "fileDecoder.csv")
    decoder = np.genfromtxt(atm_path, delimiter=",", skip_header=1)

    decoder = decoder[decoder[:, 3] == ihaze]  # downselects to the right ihaze mode
    decoder = decoder[decoder[:, 1] == altitude / 1000.0]  # downselects to the right altitude
    decoder = decoder[decoder[:, 2] == ground_range / 1000.0]  # downselects to the right ground range

    if not os.path.isfile(dir_path + "/atms/" + str(int(decoder[0, 0])) + ".bin"):
        raise IndexError("No atm file found for provided condition. Change values or use interpolation as necessary.")
    raw_data = np.fromfile(
        dir_path + "/atms/" + str(int(decoder[0, 0])) + ".bin",
        dtype=np.float32,
        count=-1,
    )

    raw_data = raw_data.reshape((1371, 5), order="F")
    raw_data[:, 1:5] = raw_data[:, 1:5] / 1e-10  # convert radiance columns to W/(sr m^2 m)

    # append wavelength as first column
    wavelength = 1e-6 * np.linspace(0.3, 14.0, 1371)
    wavelength = np.expand_dims(wavelength, axis=1)

    return np.hstack((wavelength, raw_data))


def _get_ground_range_array(max_ground_range: float) -> np.ndarray:
    """Returns an array of ground ranges that are valid in the precalculated MODTRAN database.

    :param max_ground_range:
        largest ground Range of interest (m)

    :return:
        g:
            array of ground ranges less than max_ground_range (m)
    """
    g = np.array([0.0, 100.0, 500.0])
    g = np.append(g, np.arange(1000.0, 20000.01, 1000.0))
    g = np.append(g, np.arange(22000.0, 80000.01, 2000.0))
    g = np.append(g, np.arange(85000.0, 300000.01, 5000.0))

    return g[g <= max_ground_range]


def _alt_atm_interp(
    low_alt: float,
    high_alt: float,
    altitude: float,
    ground_range: float,
    ihaze: int,
) -> np.ndarray:
    # this is an internal function for interpolating atmospheres across
    # altitudes
    low_atm = load_database_atmosphere_no_interp(low_alt, ground_range, ihaze)
    if low_alt != high_alt:
        high_atm = load_database_atmosphere_no_interp(high_alt, ground_range, ihaze)
        low_weight = 1 - ((altitude - low_alt) / (high_alt - low_alt))
        high_weight = (altitude - low_alt) / (high_alt - low_alt)
        atm = low_weight * low_atm + high_weight * high_atm
    else:
        atm = low_atm
    return atm


def load_database_atmosphere(
    altitude: float,
    ground_range: float,
    ihaze: int,
) -> np.ndarray:
    """Linear interpolation of the pre-calculated MODTRAN atmospheres.

    See the original 'load_database_atmosphere' (now commented out) for more
    details on the outputs.
    NOTE: This is experimental code.  Linear interpolation between atmospheres
    may not be a good approximation in every case!!!!


    :param altitude:
        sensor height above ground level in meters
    :param ground_range:
        distance *on the ground* between the target and sensor in meters.
        The following ground ranges are included in the database at each
        altitude until the ground range exceeds the distance to the spherical
        earth horizon: 0 100 500 1000 to 20000 in 1000 meter steps, 22000 to
        80000 in 2000 m steps, and  85000 to 300000 in 5000 meter steps.
    :param ihaze:
        MODTRAN code for visibility, valid options are ihaze = 1 (Rural
        extinction with 23 km visibility) or ihaze = 2 (Rural extinction
        with 5 km visibility)

    :return:
        atm[:,0]:
            wavelengths from .3 to 14 x 10^-6 m in 0.01x10^-6 m steps
        atm[:,1]:
            (TRANS) total transmission through the defined path.
        atm[:,2]:
            (PTH THRML) radiance component due to atmospheric emission and
            scattering received at the observer.
        atm[:,3]:
            (SURF EMIS) component of radiance due to surface emission received
            at the observer.
        atm[:,4]:
            (SOL SCAT) component of scattered solar radiance received at the
            sobserver.
        atm[:,5]:
            (GRND RFLT) is the total solar flux impingent on the ground and
            reflected directly to the sensor from the ground. (direct radiance
            + diffuse radiance) * surface reflectance
    :NOTE: units for columns 1 through 5 are in radiance W/(sr m^2 m)
    """

    # define arrays of all possible altitude and ground ranges
    altitude_array = np.array(
        [
            2,
            32.55,
            75,
            150,
            225,
            500,
            1000,
            2000,
            3000,
            4000,
            5000,
            6000,
            7000,
            8000,
            9000,
            10000,
            11000,
            12000,
            14000,
            16000,
            18000,
            20000,
        ],
    )
    ground_ranges = _get_ground_range_array(301e3)

    if altitude < min(altitude_array) or altitude > max(altitude_array):
        raise IndexError("Altitude not within range of acceptable values (2,20000)")
    if ground_range < min(ground_ranges) or ground_range > max(ground_ranges):
        raise IndexError("Ground Range not within range of acceptable values (0,300000)")
    # find the database altitudes and ground ranges that bound the values of
    # interest
    low_alt = altitude_array[altitude_array <= altitude][-1]
    high_alt = altitude_array[altitude_array >= altitude][0]
    low_range = ground_ranges[ground_ranges <= ground_range][-1]
    high_range = ground_ranges[ground_ranges >= ground_range][0]

    # first interpolate across the low and high altitudes
    # then interpolate across ground range
    atm_low_range = _alt_atm_interp(low_alt, high_alt, altitude, low_range, ihaze)
    if low_range != high_range:
        atm_high_range = _alt_atm_interp(low_alt, high_alt, altitude, high_range, ihaze)
        low_weight = 1 - ((ground_range - low_range) / (high_range - low_range))
        high_weight = (ground_range - low_range) / (high_range - low_range)
        atm = low_weight * atm_low_range + high_weight * atm_high_range
    else:
        atm = atm_low_range

    return atm
