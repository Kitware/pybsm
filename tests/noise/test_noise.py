from contextlib import AbstractContextManager
from contextlib import nullcontext as does_not_raise

import numpy as np
import pytest
from syrupy.assertion import SnapshotAssertion

from pybsm import noise


class TestNoise:
    @pytest.mark.parametrize(
        ("kernel", "expectation"),
        [
            (np.array([[0.25, 0.25], [0.25, 0.25]]), does_not_raise()),
            (np.array([[0.5], [0.5]]), does_not_raise()),
            (np.array([[1.25, 0.25], [0.25, 0.25]]), pytest.raises(ValueError, match="Kernel does not sum to 1")),
            (np.array([[-1.25, 0.25], [0.25, 0.25]]), pytest.raises(ValueError, match="Kernel does not sum to 1")),
        ],
    )
    def test_noise_gain(
        self,
        fuzzy_snapshot: SnapshotAssertion,
        kernel: np.ndarray,
        expectation: AbstractContextManager,
    ) -> None:
        """Test noise_gain against gold standard results and confirm exceptions are appropriately raised."""
        with expectation:
            output = noise.noise_gain(kernel=kernel)
            fuzzy_snapshot.assert_match(output)

    @pytest.mark.parametrize(
        ("pe_range", "bit_depth"),
        [
            (0.0, 0.0),
        ],
    )
    def test_quantization_noise_nan(self, pe_range: float, bit_depth: float) -> None:
        """Cover cases where nan occurs."""
        output = noise.quantization_noise(pe_range=pe_range, bit_depth=bit_depth)
        assert np.isnan(output)

    @pytest.mark.parametrize(
        ("pe_range", "bit_depth"),
        [
            (1.0, 0.0),
        ],
    )
    def test_quantization_noise_inf(self, pe_range: float, bit_depth: float) -> None:
        """Cover cases where inf occurs."""
        output = noise.quantization_noise(pe_range=pe_range, bit_depth=bit_depth)
        assert np.isinf(output)

    @pytest.mark.parametrize(
        ("pe_range", "bit_depth"),
        [
            (0.0, 1.0),
            (1.0, 1.0),
        ],
    )
    def test_quantization_noise(self, pe_range: float, bit_depth: float, fuzzy_snapshot: SnapshotAssertion) -> None:
        """Test quantization_noise with normal inputs and expected outputs."""
        output = noise.quantization_noise(pe_range=pe_range, bit_depth=bit_depth)
        fuzzy_snapshot.assert_match(output)
