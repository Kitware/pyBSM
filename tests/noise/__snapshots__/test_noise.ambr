# serializer version: 1
# name: TestNoise.test_noise_gain_valid[kernel0-expectation0]
  0.5
# ---
# name: TestNoise.test_noise_gain_valid[kernel1-expectation1]
  0.7071067811865476
# ---
